feat: Noir subrepo.

charlielye opened the pull request:
Finally.

Builds the now subrepo'd noir repository. At present this just builds nargo for x86 and arm, meaning our sandbox install script can now provide a wrapper `aztec-nargo` that is guaranteed to always be exactly what we want, regardless of any modifications to macros, compilers, packages, or whatever. Decouples aztec from being dependent on Noirs deployment pipeline, while still providing the ability to push changes back to Noir.

Still need to do work to build npm packages and portal them into yarn-project. Will do in separate PR.

* Removes old benchmarking stuff and commented canary stuff from pipeline.
* Adds nargo x86 and arm builds, outputing a multiarch image that also runs efficiently on macs.
* Removes a load of ARCH specific hack stuff in build-system, in favour of all cache image URI's now just have their arch appended to their tag. The arch is determined by the arch of the build system. This required a small "hack(?)" whereby we will fall back from an unfound arch (arm) to x86, as we build some of our arm images from previously build x86 images, but i anticipate that'll change at some point.
* Remove some project scripts that bleed the build-system abstraction (e.g. deploy_docker.sh).
* Introduces `[ci dry-deploy]` commit message command for doing dry run of deploys. We no longer have conditional workflow filters for deployment to enable this. I.e. there is always deploy jobs, they just noop out asap. If dry-deploy is enabled, the deploy jobs will run to "completion" but doesn't actually execute pushes to dockerhub, and just runs `npm publish` in dry-run mode.
* Gets rid of `VERSION` files and the sanity check is it got in the way of the above, and I don't even remember why it was there.
* `build-system` can now launch arm spot instances. This can be requested e.g. `cond_spot_run_build noir 32 arm64`.
* Added a script to help bootstrap build-instances in case we need to create new AMI's again in future (I had to create and arm one).
* Deleted some build-system scripts I couldn't see used anywhere.
* `deploy_dockerhub` script now takes list of arches and does the manifest generation itself.
* Introduces a `should_deploy` script which enables the early out of deployment jobs. At present this causes exit if there is no COMMIT_TAG. In future it'll want to run the deploy steps if BRANCH is master as well.


Follow up PRs will:
* Build NPM packages and portal them into yarn-project.
* Modify sandbox install script to pull noir image, install `aztec-nargo` wrapper.
* Update vscode plugin to allow selection between global `nargo` or `aztec-nargo` if both found.
* Someone will do some magic to use the nargo container in github codespaces so users can press `.` and play.

charlielye committed:
git subrepo clone git@github.com:noir-lang/noir

subrepo:
  subdir:   "noir"
  merged:   "d27db332f"
upstream:
  origin:   "git@github.com:noir-lang/noir"
  branch:   "master"
  commit:   "d27db332f"
git-subrepo:
  version:  "0.4.6"
  origin:   "https://github.com/ingydotnet/git-subrepo"
  commit:   "110b9eb"

charlielye committed:
Build noir.

ludamad commented:
![image](https://github.com/AztecProtocol/aztec-packages/assets/163993/d4a17078-6e22-4d4a-b40a-61ce702453ec)

charlielye commented:
![](https://github.com/AztecProtocol/aztec-packages/assets/5764343/44e1ca61-f933-4e63-84e2-900fccd6fbdd)

charlielye committed:
Arm support.

charlielye committed:
[ci rebuild noir-x86_64] [ci rebuild noir-arm64]

charlielye committed:
[ci rebuild noir]

charlielye committed:
[ci rebuild noir]

charlielye committed:
[ci rebuild noir]

charlielye committed:
[ci rebuild noir]

charlielye committed:
[ci rebuild noir]

charlielye committed:
[ci rebuild noir]

charlielye committed:
[ci rebuild noir]

charlielye committed:
git subrepo pull (merge) noir

subrepo:
  subdir:   "noir"
  merged:   "816deeb60"
upstream:
  origin:   "git@github.com:noir-lang/noir"
  branch:   "master"
  commit:   "63d414c06"
git-subrepo:
  version:  "0.4.6"
  origin:   "https://github.com/ingydotnet/git-subrepo"
  commit:   "110b9eb"

charlielye committed:
[ci rebuild noir]

charlielye committed:
Noir deploy

charlielye committed:
git subrepo commit (merge) noir

subrepo:
  subdir:   "noir"
  merged:   "00af10514"
upstream:
  origin:   "git@github.com:noir-lang/noir"
  branch:   "master"
  commit:   "251b930de"
git-subrepo:
  version:  "0.4.6"
  origin:   "https://github.com/ingydotnet/git-subrepo"
  commit:   "110b9eb"

charlielye committed:
Move to machine class to get docker

charlielye committed:
[ci rebuild noir]

charlielye committed:
Alpine release build of noir.

charlielye committed:
Remove all the ARCH hacks from build system. Simplify deploy_dockerhub.

charlielye committed:
Remove unused save log stuff.

charlielye committed:
Fixes. Can trigger dry deploy via cci.

charlielye committed:
Try again. Now there is workflow distinction for deploy, but we early out, unless dry deploy is set.

charlielye committed:
fix

charlielye committed:
fix

charlielye committed:
fix

charlielye committed:
fix

charlielye committed:
fix

charlielye committed:
fix

charlielye committed:
fix

charlielye committed:
Fix version number.

charlielye committed:
fix

charlielye committed:
fix

charlielye committed:
fix

charlielye committed:
fix

charlielye committed:
fix

charlielye committed:
fix

charlielye committed:
Remove the VERSION files and sanity check test in setup-env.

charlielye committed:
fix

charlielye committed:
fix

charlielye committed:
fix

charlielye committed:
fix

charlielye committed:
[ci dry-deploy]

charlielye committed:
[ci dry-deploy]

charlielye committed:
Merge remote-tracking branch 'origin/master' into cl/noir-subrepo

charlielye committed:
Merge branch 'master' into cl/noir-subrepo

charlielye committed:
fix

charlielye committed:
[ci dry-deploy]

charlielye committed:
fix [ci dry-deploy]

spypsy commented on .circleci/config.yml:
```diff
@@ -993,92 +960,61 @@ jobs:
+          command: |
+            [ -n "$COMMIT_TAG" ] || exit 0
+            deploy_dockerhub noir x86_64,arm64
+            deploy_dockerhub pxe x86_64,arm64
```
pxe image has been removed as it can now be started from the aztec-sandbox image with `MODE=pxe`

charlielye committed:
fix [ci dry-deploy]

spypsy commented on build_manifest.yml:
```diff
@@ -162,3 +162,9 @@ docs:
+
+noir:
+  buildDir: noir
+  arch:
```
Is this used somewhere?

charlielye replied:
Actually no. Well spotted. Intention is to replace some of the hardcoded stuff in config.yml with a manifest query. e.g.
```
create_ecr_manifest aztec-sandbox x86_64,arm64
create_ecr_manifest cli x86_64,arm64
```           
becomes
```
create_ecr_manifest aztec-sandbox
create_ecr_manifest cli
```

charlielye replied:
And ultimately I'm still hopeful of fully code generating config.yml from the build manifest.

charlielye committed:
Address review comments. [ci rebuild]
